geomesa-gt/geomesa-gt-partitioning/src/main/scala/org/locationtech/geomesa/gt/partition/postgis/dialect/PartitionedPostgisDialect.scala

```diff
@@ -0,0 +1,188 @@
+/***********************************************************************
+ * Copyright (c) 2013-2022 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.gt.partition.postgis.dialect
+
+import com.typesafe.scalalogging.StrictLogging
+import org.geotools.data.postgis.PostGISDialect
+import org.geotools.jdbc.JDBCDataStore
+import org.locationtech.geomesa.gt.partition.postgis.dialect.functions.{LogCleaner, TruncateToPartition, TruncateToTenMinutes}
+import org.locationtech.geomesa.gt.partition.postgis.dialect.procedures._
+import org.locationtech.geomesa.gt.partition.postgis.dialect.tables._
+import org.locationtech.geomesa.gt.partition.postgis.dialect.triggers.{DeleteTrigger, InsertTrigger, UpdateTrigger, WriteAheadTrigger}
+import org.locationtech.geomesa.utils.geotools.{Conversions, SimpleFeatureTypes}
+import org.opengis.feature.simple.SimpleFeatureType
+
+import java.sql.{Connection, DatabaseMetaData}
+import scala.collection.mutable.ArrayBuffer
+
+class PartitionedPostgisDialect(store: JDBCDataStore) extends PostGISDialect(store) with StrictLogging {
+
+  // order of calls from JDBCDataStore during create schema:
+  //  encodeCreateTable
+  //  encodeTableName
+  //  encodePrimaryKey
+  //  encodeColumnName
+  //  encodeColumnType
+  //  encodePostColumnCreateTable
+  //  encodePostCreateTable
+  //  postCreateTable
+
+  // order of calls during remove schema:
+  //  preDropTable
+  //  "DROP TABLE" + encodeTableName
+  //  postDropTable
+
+  // state for checking when we want to use the write_ahead table in place of the main view
+  private val replaceTableName = new ThreadLocal[Boolean]() {
+    override def initialValue(): Boolean = false
+  }
+
+  // filter out the partition tables from exposed feature types
+  override def getDesiredTablesType: Array[String] = Array("VIEW")
+
+  override def encodeTableName(raw: String, sql: StringBuffer): Unit = {
+    sql.append(s""""${if (replaceTableName.get) { raw + WriteAheadTableSuffix } else { raw }}"""")
+    replaceTableName.remove()
+  }
+
+  override def encodePrimaryKey(column: String, sql: StringBuffer): Unit = {
+    encodeColumnName(null, column, sql)
+    // make our primary key a string instead of the default integer
+    sql.append(" character varying NOT NULL")
+  }
+
+  override def postCreateTable(schemaName: String, sft: SimpleFeatureType, cx: Connection): Unit = {
+    val info = TypeInfo(schemaName, sft)
+
+    super.postCreateTable(schemaName, sft, cx)
+
+    implicit val ex: ExecutionContext = new ExecutionContext(cx)
+    try {
+      val commands = ArrayBuffer.empty[Sql]
+      commands += WriteAheadTable
```

Comment:
Does the order matter?  And seems like a Seq() of these might read better?  I don't know what syntactically easiest to read here...

Reply:
The order does matter. I mainly did it this way to avoid having to ++ around the awkward optional procedure, but I don't mind changing it if you'd prefer.

Reply:
*shrugs*  It is fine.  I think all the syntax for this will be weird.